integreat_cms/deepl_api/utils.py

```diff
@@ -48,6 +48,42 @@ def check_availability(request, language_slug):
             and language_slug in deepl_config.supported_target_languages
         )
 
+    def get_target_language_key(self, request, target_language):
+        """
+        This function decides the correct target language key
+
+        :param request: the transmitted request
+        :type request: ~django.http.HttpRequest
+
+        :param target_language: the target language
+        :type target_language: ~integreat_cms.cms.models.abstract_content_translation.AbstractContentTranslation
+
+        :return: target_language_key which is 2 characters long for all languages except English and Portugese where the BCP code is transmitted
+        :rtype: str
+        """
+        # For some languages, the DeepL client expects the BCP tag instead of the short language code
+        if target_language.slug in ("en", "pt"):
+            # Check if BCP 47 tag is five characters as expected and contains a "-"
+            if len(target_language.bcp47_tag) == 5 and "-" in target_language.bcp47_tag:
+                target_language_key = target_language.bcp47_tag
+            # If not, check if instead of the BCP47 code the slug was entered
+            elif len(target_language.bcp47_tag) == 2:
+                # Set Great Britain as default for English
+                if target_language.bcp47_tag == "en":
+                    target_language_key = target_language.slug + "-GB"
+                # Set Portugal as default for Portugese
+                elif target_language.bcp47_tag == "pt":
+                    target_language_key = target_language.slug + "-PT"
+            else:
+                messages.error(
+                    request,
+                    _("Your bcp code for the selected language seems to be incorrect."),
+                )
+                target_language_key = None
+        else:
+            target_language_key = target_language.slug
+        return target_language_key
+
```

Comment:
```suggestion
        if target_language.slug not in ["en", "pt"]:
            return target_language.slug

        if len(target_language.bcp47_tag) == 5 and "-" in target_language.bcp47_tag:
            return target_language.bcp47_tag

        defaults = {
            "en": "en-GB",
            "pt": "pt-PT",
        }
        return defaults.setdefault(target_language.slug, None)
```
I think this might be a little better readable, and is easily extendable should the need arise.

I'd then move the error message into the `deepl_translation`  function (see below)

Reply:
@charludo I like your simpler version of the code!
However, I think in this case we don't even need the hard-coded fallback mechanism - I would suggest to handle this case already at the "check availability"-stage and just not enable DeepL translations for en or pt if the bcp tag is faulty. I think this is too much of an edge case to provide detailed user feedback here. Also, the assumption that a valid bcp tag always has 5 characters is wrong...

```suggestion
        deepl_config = apps.get_app_config("deepl_api")
        for code in [target_language.slug, target_language.bcp47_tag]:
            if code in deepl_config.supported_source_languages:
                return code
        return None

```

My suggestion would be to make this completely language agnostic and just remove the `[:2]` here: https://github.com/digitalfabrik/integreat-cms/blob/e90baf5755c9ef18e7b943c6308c9c37b8ec1426/integreat_cms/deepl_api/apps.py#L51

This would mean we would store the bcp tag in the `supported_source_languages` config and could then check whether either the slug or the bcp tag are supported.

This would also require the `check_availability()` function to be rewritten to also check for the bcp tag if the slug is not valid.

Reply:
Is there a preferred way on how I should go about it? Would it be better if I added a second parameter for the bcp code or should I rather change the language_slug parameter to a language parameter, that contains both slug and bcp47 code as attributes?

Reply:
I think one single language parameter makes the most sense...